### PR TITLE
Fixed requesting binaries

### PR DESCRIPTION
**Pull request by obiwac**
This is my first time creating a pull request, so I don't really know how all this works but... I've modified `resp_callback` in `src/requests.c` to work with binaries instead of just plain text.

### CODE REVIEW

**Comment by mossberg** [comment]
Sorry for missing this PR! I'll take a look. Thanks so much for submitting it.

File: src/requests.c [hunk 1]
```diff
@@ -108,22 +110,22 @@ static size_t resp_callback(char *content, size_t size, size_t nmemb,
...
 
-    /* create NULL terminated version of `content' */
-    char *responsetext = strndup(content, real_size + 1);
-    if (responsetext == NULL)
-        return -1;
-
-    strncat(userdata->text, responsetext, real_size);
+	/* concatenate userdata->text with the response content */
+	int i;
+	for (i = original_userdata_size; i < original_userdata_size + real_size; i++) {
```
Comment (mossberg) [inline src/requests.c]: could we use memcpy here?

File: src/requests.c [hunk 2]
```diff
@@ -24,7 +24,9 @@
  * THE SOFTWARE.
  */
 
-#include "requests.h"
+#ifndef __OMIT_INCLUSION_OF_REQUESTS_HEADER__
```
Comment (mossberg) [inline src/requests.c]: what exactly is the motivation for this change?

Comment (obiwac) [reply src/requests.c]: Oh sorry that was a mistake on my part. I accidentally left it while debugging.

**Comment by obiwac** [comment]
After 3 years of tireless Google-searching, I now finally understand how pull requests work ;)

File: src/requests.c [hunk 3]
```diff
@@ -117,22 +117,16 @@ static size_t resp_callback(char *content, size_t size, size_t nmemb,
...
-    /* create NULL terminated version of `content' */
-    char *responsetext = strndup(content, real_size + 1);
-    if (responsetext == NULL)
-        return -1;
-
-    strncat(userdata->text, responsetext, real_size);
-
-    free(responsetext);
+	/* concatenate userdata->text with the response content */
+	memcpy(userdata->text + original_userdata_size, content, real_size);
```
Comment (offlinemark) [inline src/requests.c:129]: Since we're not using strncat anymore (presumably to support binary data), we need to remember to NULL-terminate the string manually. I believe it would look like this:

```suggestion
	memcpy(userdata->text + original_userdata_size, content, real_size);
	userdata->text[original_userdata_size + real_size] = '\0';
```

File: src/requests.c [hunk 4]
```diff
@@ -34,7 +34,7 @@ static char *user_agent(void);
 static int check_ok(long code);
 static CURLcode requests_pt(req_t *req, char *url, char *data,
                             char **custom_hdrv, int custom_hdrc, int put_flag);
-static int hdrv_append(char ***hdrv, int *hdrc, char *new);
+static int hdrv_append(char ***hdrv, int *hdrc, char *_new);
```
Comment (offlinemark) [inline src/requests.c:37]: Curious why these changes are necessary? Did you try to compile this with a C++ compiler?

File: src/requests.c [hunk 5]
```diff
@@ -351,10 +345,10 @@ static CURLcode requests_pt(req_t *req, char *url, char *data,
     } else {
         /* content length header defaults to -1, which causes request to fail
            sometimes, so we need to manually set it to 0 */
-        char *cl_header = "Content-Length: 0";
+        char *cl_header = (char*) "Content-Length: 0";
```
Comment (offlinemark) [inline src/requests.c:348]: Also interested in what prompted this change?

**Review by offlinemark — COMMENTED** [review]
Welcome back @obiwac! Thanks for the changes. I have to say under normal circumstances, I'd probably work with you to shape this up more, but since I don't use this at all (and don't have much time for this), and you clearly do, I'm inclined to merge it. Can you tell me what you're using it for, and confirm that these changes are working for your use case?

File: src/requests.c (hunk 3 above)
Comment (obiwac) [reply src/requests.c:129]: Indeed, that seems a bit better :)

File: src/requests.c (hunk 5 above)
Comment (obiwac) [reply src/requests.c:349]: No clue.

**Review by obiwac — COMMENTED** [review]
Currently tryna clean things up.

**Comment by obiwac** [comment]
Ok I think things should be good now. I replaced `asprintf` so that this can compile with non-GNU/BSD libc, and everything compiles and runs just fine on my machine.

As for what I use this project for, it's really mostly for random projects for which I need a C equivalent to Python's requests library.

**Comment by offlinemark** [comment]
Cool. Will merge since it looks ok and most importantly seems to fix the truncation issues with binary files (though I never tested - taking your word for it). I can see us continuing to do cleanups in the future (revert the new -> _new stuff, remove the unnecessary casts, etc) but that can be for later. Thanks 👍

**Comment by obiwac** [comment]
Thank you!
